server/src/main/java/org/candlepin/resource/util/ResolverUtil.java

```diff
@@ -78,35 +80,57 @@ public Owner resolveOwner(Owner owner) {
         return owner;
     }
 
-    public Product resolveProduct(Owner owner, Product product) {
-        String id = null;
-        if (product != null) {
-            id = product.getId();
-        }
-        return resolveProduct(owner, id);
+    public Product resolveProduct(Owner owner, String productId, boolean allowNull) {
+        Product product = new Product();
+        product.setId(productId);
+        return resolveProduct(owner, product, allowNull);
     }
 
-    public Product resolveProduct(Owner owner, String productId) {
-        if (productId == null) {
-            throw new BadRequestException(
-                    i18n.tr("No product specified, or product lacks identifying information"));
+    public ProductData resolveProduct(Owner owner, ProductData product, boolean allowNull) {
+        Product argProduct = new Product();
+        ProductData result = null;
+        if (product != null) {
+            argProduct.setId(product.getId());
+            argProduct.setUuid(product.getUuid());
         }
-
-        // TODO: Maybe add UUID resolution as well?
-        return findProduct(owner, productId);
+        if (product != null) {
+            result = new ProductData(resolveProduct(owner, argProduct, allowNull));
+        }
+        return result;
     }
 
-    public Product findProduct(Owner owner, String productId) {
-        Product product = this.ownerProductCurator.getProductById(owner, productId);
-
-        if (product == null) {
-            throw new NotFoundException(i18n.tr(
-                "Unable to find a product with the ID \"{0}\" for owner \"{1}\"",
-                productId, owner.getKey()
+    public Product resolveProduct(Owner owner, Product product, boolean allowNull) {
+        Product result = null;
+        if (product != null) {
+            if (product.getUuid() != null) {
+                result = this.productCurator.find(product.getUuid());
+                if (result == null) {
+                    throw new NotFoundException(i18n.tr(
+                        "Unable to find a product with the UUID \"{0}\"", product.getUuid()
+                    ));
+                }
+            }
+            else if (product.getId() != null) {
+                result = this.ownerProductCurator.getProductById(owner, product.getId());
+                if (result == null) {
+                    throw new NotFoundException(i18n.tr(
+                        "Unable to find a product with the ID \"{0}\" for owner \"{1}\"",
+                            product.getId(), owner.getKey()
+                    ));
+                }
+            }
+            else {
+                throw new BadRequestException(i18n.tr(
+                   "No product specified, or product lacks identifying information"
+                ));
+            }
+        }
+        else if (!allowNull) {
```

Comment:
Something about this kinda irks me. Entity resolution wasn't intended to be drop-in input validation. If we're resolving a product, the assumption here is that we should always have something, or it should be completely passthrough.